File: server/apps/nexotype/schemas/knowledge_graph/bioactivity_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class BioActivityCreate(BaseModel):
    """Schema for creating a new bioactivity"""
    asset_id: int
    pathway_id: int
    activity_type: str


class BioActivityUpdate(BaseModel):
    """Schema for updating a bioactivity"""
    asset_id: int = None
    pathway_id: int = None
    activity_type: str = None


# ==========================================
# Response Schemas
# ==========================================

class BioActivityDetail(BaseModel):
    """Schema for bioactivity details"""
    id: int
    asset_id: int
    pathway_id: int
    activity_type: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class BioActivityListResponse(BaseModel):
    """Response schema for listing bioactivities"""
    success: bool
    data: list[BioActivityDetail] | None = None
    count: int = 0
    error: str | None = None


class BioActivityResponse(BaseModel):
    """Response schema for single bioactivity operations"""
    success: bool
    data: BioActivityDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/knowledge_graph/bioactivity.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

export const ActivityTypeEnum = z.enum(['Activator', 'Senolytic', 'NAD+ Booster']);

// ==========================================
// Options (for Select dropdowns)
// ==========================================

/** BioActivity.activity_type */
export const ACTIVITY_TYPE_OPTIONS = [
  { label: 'Activator', value: 'Activator' },
  { label: 'Senolytic', value: 'Senolytic' },
  { label: 'NAD+ Booster', value: 'NAD+ Booster' },
] as const;

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const BioactivitySchema = z.object({
  id: z.number(),
  asset_id: z.number().int(),
  pathway_id: z.number().int(),
  activity_type: ActivityTypeEnum,
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateBioactivitySchema = z.object({
  asset_id: z.number().int(),
  pathway_id: z.number().int(),
  activity_type: ActivityTypeEnum,
});

export const UpdateBioactivitySchema = z.object({
  asset_id: z.number().int().optional(),
  pathway_id: z.number().int().optional(),
  activity_type: ActivityTypeEnum.optional(),
});

// ==========================================
// Types
// ==========================================

export type Bioactivity = z.infer<typeof BioactivitySchema>;
export type CreateBioactivity = z.infer<typeof CreateBioactivitySchema>;
export type UpdateBioactivity = z.infer<typeof UpdateBioactivitySchema>;
export type ActivityType = z.infer<typeof ActivityTypeEnum>;

// ==========================================
// Response Types
// ==========================================

export type BioactivityResponse = {
  success: boolean;
  data?: Bioactivity;
  error?: string;
};

export type BioactivitiesResponse = {
  success: boolean;
  data?: Bioactivity[];
  count?: number;
  error?: string;
};
